cmd/lotus-shed/gas-estimation.go

```diff
@@ -0,0 +1,178 @@
+// the epoch you migrated at, the network version you migrated to, and a message hash. You will be able to replay any
+// message from between the migration epoch, and where your node originally synced to. Note: You may run into issues
+// with nonces, or state that changed between the epoch you migrated at, and when the message was originally processed.
+// This can be avoided by replaying messages from close to the migration epoch, or circumvented by using a custom
+// FVM bundle.
+var gasTraceCmd = &cli.Command{
+	Name:        "trace-gas",
+	Description: "replay a message on the specified stateRoot and network version to get an execution trace",
+	ArgsUsage:   "[migratedStateRootCid migrationEpoch networkVersion messageHash]",
+	Flags: []cli.Flag{
+		&cli.StringFlag{
+			Name:  "repo",
+			Value: "~/.lotus",
+		},
+	},
+	Action: func(cctx *cli.Context) error {
+		ctx := context.TODO()
+
+		if cctx.NArg() != 4 {
+			return lcli.IncorrectNumArgs(cctx)
+		}
+
+		stateRootCid, err := cid.Decode(cctx.Args().Get(0))
+		if err != nil {
+			return fmt.Errorf("failed to parse input: %w", err)
+		}
+
+		epoch, err := strconv.ParseInt(cctx.Args().Get(1), 10, 64)
+		if err != nil {
+			return fmt.Errorf("failed to parse input: %w", err)
+		}
+
+		nv, err := strconv.ParseInt(cctx.Args().Get(2), 10, 32)
+		if err != nil {
+			return fmt.Errorf("failed to parse input: %w", err)
+		}
+
+		messageCid, err := cid.Decode(cctx.Args().Get(3))
+		if err != nil {
+			return fmt.Errorf("failed to parse input: %w", err)
+		}
+
+		fsrepo, err := repo.NewFS(cctx.String("repo"))
+		if err != nil {
+			return err
+		}
+
+		lkrepo, err := fsrepo.Lock(repo.FullNode)
+		if err != nil {
+			return err
+		}
+
+		defer lkrepo.Close() //nolint:errcheck
+
+		bs, err := lkrepo.Blockstore(ctx, repo.UniversalBlockstore)
+		if err != nil {
+			return fmt.Errorf("failed to open blockstore: %w", err)
+		}
+
+		defer func() {
+			if c, ok := bs.(io.Closer); ok {
+				if err := c.Close(); err != nil {
+					log.Warnf("failed to close blockstore: %s", err)
+				}
+			}
+		}()
+
+		mds, err := lkrepo.Datastore(context.Background(), "/metadata")
+		if err != nil {
+			return err
+		}
+
+		dcs := build.DrandConfigSchedule()
+		shd := beacon.Schedule{}
+		for _, dc := range dcs {
+			bc, err := drand.NewDrandBeacon(1598306400, build.BlockDelaySecs, nil, dc.Config)
+			if err != nil {
+				return xerrors.Errorf("creating drand beacon: %w", err)
+			}
+			shd = append(shd, beacon.BeaconPoint{Start: dc.Start, Beacon: bc})
+		}
+		cs := store.NewChainStore(bs, bs, mds, filcns.Weight, nil)
+		defer cs.Close() //nolint:errcheck
+
+		sm, err := stmgr.NewStateManager(cs, filcns.NewTipSetExecutor(), vm.Syscalls(ffiwrapper.ProofVerifier), filcns.DefaultUpgradeSchedule(), shd)
+		if err != nil {
+			return err
+		}
+
+		msg, err := cs.GetMessage(ctx, messageCid)
+		if err != nil {
+			return err
+		}
+
+		// Set to block limit so message will not run out of gas
+		msg.GasLimit = 10_000_000_000
```

Comment:
1) A solution modeled on `Call` instead of `CallWithGas` might be what you want here -- that will basically never run out of gas.
2) If this line is staying, use the constant for block gas limit here.

Reply:
Gonna use a constant here.